### PR TITLE
Added cr7ial support

### PR DESCRIPTION
**Pull request by dniewinski**
Added a cr7ia support library and a urdf and moveit configuration for a cr7ial.  Tested on a real arm and in roboguide

### CODE REVIEW

**Comment by gavanderhoorn** [comment]
Thanks again for the PR.

Unfortunately, Travis is failing:

```
FAILURE:
[/root/catkin_ws/src/fanuc_experimental/fanuc_cr7ia_support/tests/roslaunch_test.xml]:
	while processing /root/catkin_ws/src/fanuc_experimental/fanuc_cr7ia_support/launch/load_cr7ia.launch:
Invalid roslaunch XML syntax: [Errno 2] No such file or directory: u'/root/catkin_ws/src/fanuc_experimental/fanuc_cr7ia_support/launch/load_cr7ia.launch'
```

It should probably be `load_cr7ial.launch`?

File: fanuc_cr7ia_support/package.xml
Comment (gavanderhoorn) [inline fanuc_cr7ia_support/package.xml:17]: Just making sure: have you verified this? What did you choose in the initial robot setup?

File: fanuc_cr7ia_support/package.xml
Comment (gavanderhoorn) [inline fanuc_cr7ia_support/package.xml]: Could you please be exact about which version of the manual you used? Fanuc documents are versioned, so I'd like this to be as specific as possible. It should be something like `B-NNNNNLL/VV` (with `NNNNN` the doc id, `LL` the language and `VV` the revision).

File: fanuc_cr7ia_support/package.xml
Comment (gavanderhoorn) [inline fanuc_cr7ia_support/package.xml]: I've added myself as maintainer, as we do that for all ROS-I packages.

Are you actually planning to maintain this? I'd very much welcome that, but it is a committment.

File: fanuc_cr7ial_moveit_config/.setup_assistant
Comment (gavanderhoorn) [inline fanuc_cr7ial_moveit_config/.setup_assistant:5]: I've added this to make MoveIt enable Jade+ xacro on Indigo.

File: fanuc_cr7ial_moveit_config/config/fanuc_cr7ial.srdf
Comment (gavanderhoorn) [inline fanuc_cr7ial_moveit_config/config/fanuc_cr7ial.srdf]: This `group_state` seems rather specific to your setup on the Ridgeback, could you remove it from this config?

File: fanuc_cr7ial_moveit_config/config/fanuc_cr7ial.srdf
Comment (gavanderhoorn) [inline fanuc_cr7ial_moveit_config/config/fanuc_cr7ial.srdf]: I'm not a native speaker, but I believe the plural of `zero` is `zeros`. `zeroes` would be the verb.

File: fanuc_cr7ial_moveit_config/config/fanuc_cr7ial.srdf
Comment (gavanderhoorn) [inline fanuc_cr7ial_moveit_config/config/fanuc_cr7ial.srdf]: Could I ask you to change the `name` of the `virtual_joint` to `FixedBase`? That is a ROS-I convention.

Additionally, the `parent_frame` would be `world`.

File: fanuc_cr7ial_moveit_config/config/kinematics.yaml
Comment (gavanderhoorn) [inline fanuc_cr7ial_moveit_config/config/kinematics.yaml]: I would actually recommend you use the `trac_ik` plugin for this setup.

File: fanuc_cr7ial_moveit_config/config/ompl_planning.yaml
Comment (gavanderhoorn) [inline fanuc_cr7ial_moveit_config/config/ompl_planning.yaml]: Could you please add a `default_planner_config` entry here? See [fanuc_lrmate200ib_moveit_config](https://github.com/ros-industrial/fanuc_experimental/blob/ef1fc744b23a1fc3cf71db01b6e9dcd62b8b690c/fanuc_lrmate200ib_moveit_config/config/ompl_planning.yaml#L56) for an example.

File: fanuc_cr7ial_moveit_config/package.xml
Comment (gavanderhoorn) [inline fanuc_cr7ial_moveit_config/package.xml]: See earlier comment about `maintainer` in `fanuc_cr7ia_support`.

**Review by gavanderhoorn — CHANGES_REQUESTED** [review]
I've pushed some fixup commits already for the minor things. There are some others I've commented on here, could you take a look?

Thanks again for taking the time to contribute this.

File: fanuc_cr7ial_moveit_config/config/fanuc_cr7ial.srdf
Comment (dniewinski) [reply fanuc_cr7ial_moveit_config/config/fanuc_cr7ial.srdf]: Yup.  English is a wonderful language

File: fanuc_cr7ial_moveit_config/config/fanuc_cr7ial.srdf
Comment (dniewinski) [reply fanuc_cr7ial_moveit_config/config/fanuc_cr7ial.srdf]: I believe that is standard shipping position on this arm.  Took it from the default ship program

**Comment by gavanderhoorn** [comment]
@dniewinski: I think you just reverted *all* my changes. Was that intentional?

**Comment by dniewinski** [comment]
@gavanderhoorn That was not intentional.  I had the tests fixed up last night and forgot to push.  Pushed right when I got in without checking first.  Sorry about that.  Feel free to blow mine away, they were very minor

**Comment by gavanderhoorn** [comment]
I'll see if I can rebase mine.

Would be nice if you could look at my remaining comments in the mean time.

File: fanuc_cr7ial_moveit_config/config/fanuc_cr7ial.srdf
Comment (dniewinski) [reply fanuc_cr7ial_moveit_config/config/fanuc_cr7ial.srdf]: If this is the convention, not a problem.  I'm not a fan of the mounting point being called "world" because a robot's world isn't necessarily there (ie. fanuc), but that is just a personal opinion

File: fanuc_cr7ial_moveit_config/config/ompl_planning.yaml
Comment (dniewinski) [reply fanuc_cr7ial_moveit_config/config/ompl_planning.yaml]: Not a problem

**Comment by dniewinski** [comment]
Welcome back changes :)  Sorry again about that

File: fanuc_cr7ial_moveit_config/config/fanuc_cr7ial.srdf
Comment (gavanderhoorn) [reply fanuc_cr7ial_moveit_config/config/fanuc_cr7ial.srdf]: It's really just a convention, and I don't expect people to re-use MoveIt configs as-is anyway, apart from some initial testing maybe (who has a single robot in complete isolation to plan for).

The "world" frame that you are referring to is what Fanuc calls the *World Frame*, which is modelled with the `base` frame in the urdf.

File: fanuc_cr7ia_support/package.xml
Comment (dniewinski) [reply fanuc_cr7ia_support/package.xml]: I can maintain, but you would likely be better-suited.  I'll just contribute

File: fanuc_cr7ia_support/urdf/cr7ial_macro.xacro
Comment (gavanderhoorn) [inline fanuc_cr7ia_support/urdf/cr7ial_macro.xacro]: Not sure this should've been part of this commit?

Is it still necessary on your machine? Jade+ xacro should already have this defined.

Comment (dniewinski) [reply fanuc_cr7ia_support/urdf/cr7ial_macro.xacro]: I'm using Indigo, so I need it there for testing

Comment (gavanderhoorn) [reply fanuc_cr7ia_support/urdf/cr7ial_macro.xacro]: If you start `xacro` with the `--inorder` command line option (which I added: 461862602e5f460232e6cc17d72d891df0132b4d, and the `load_cr7ial.launch` already had) then it uses Jade+ xacro. Does that not work for you?

File: fanuc_cr7ial_moveit_config/config/fanuc_cr7ial.srdf
Comment (dniewinski) [reply fanuc_cr7ial_moveit_config/config/fanuc_cr7ial.srdf]: I'll remove this

File: fanuc_cr7ial_moveit_config/config/kinematics.yaml
Comment (dniewinski) [reply fanuc_cr7ial_moveit_config/config/kinematics.yaml]: Just curious, but what are the differences between the two?  I've had decent success with KDL so far

Comment (gavanderhoorn) [reply fanuc_cr7ial_moveit_config/config/kinematics.yaml]: Please see [traclabs/trac_ik](https://bitbucket.org/traclabs/trac_ik).

Comment (dniewinski) [reply fanuc_cr7ial_moveit_config/config/kinematics.yaml]: In my very minimal testing, KDL seems less likely to switch arm configurations than trac, or when it does, they are more minor swings

Comment (gavanderhoorn) [reply fanuc_cr7ial_moveit_config/config/kinematics.yaml]: You are probably using the `Speed` `solve_type`. You'll want to configure `Distance`. See the `trac_ik` docs.

File: fanuc_cr7ia_support/package.xml
Comment (dniewinski) [reply fanuc_cr7ia_support/package.xml:17]: Verifying this now.  J3 limits on Fanuc change depending on J2, and since it's linked-arm, the J3 value is also tied to J2.  Is there any way to have the limits of J3 updated based on the current value of J2?  Unless the limits are set small enough that they are always within the actual limits, there will always be cases where you can plan outside of actual limits

Comment (gavanderhoorn) [reply fanuc_cr7ia_support/package.xml:17]: re: limits based on J2-J3: true, but *normal range* only refers to the range of motion of J1 in this case.

File: fanuc_cr7ia_support/package.xml
Comment (dniewinski) [reply fanuc_cr7ia_support/package.xml]: Updated

File: fanuc_cr7ia_support/package.xml
Comment (dniewinski) [reply fanuc_cr7ia_support/package.xml:17]: Limits have been update as of manual.  I don't see any reference to max joint speeds in the manual though

Comment (gavanderhoorn) [reply fanuc_cr7ia_support/package.xml:17]: No, Fanuc doesn't do that for their CR range. Where'd you get the joint speed limits that you have in there now?

Comment (gavanderhoorn) [reply fanuc_cr7ia_support/package.xml:17]: Could you get me a copy of all `sy*.va` files on the controller? Either simulated or real, doesn't matter. A zip file to my email would be very much appreciated.

FTP is probably easiest.

Comment (dniewinski) [reply fanuc_cr7ia_support/package.xml:17]: I will get a backup to you shortly

**Comment by dniewinski** [comment]
I think I got all of the requested changes

**Comment by dniewinski** [comment]
Just following up.  Is there anything left to be done with this before it is merged?

**Comment by gavanderhoorn** [comment]
I just want to test it under Indigo before I merge.

**Comment by gavanderhoorn** [comment]
One more thing btw: the joint position limits for J2 and J3 that you set in b80aab5 seem to be different from the ones in the system variables. Where did you find those values?

**Comment by dniewinski** [comment]
The joint limits in the controller won't match the URDF because their kinematics are different. When you move J2, the J3 link should also move, but it doesn't.  This means that the value of that joint won't match, same with the limits.  Since the URDF is not setup as a linked-arm, it makes the J3 limits easier to do, but they won't match what the controller "thinks" its limits are.  As far as J2, the limits I have set are from the manual, not from my machine, just in case my robot is different

**Comment by gavanderhoorn** [comment]
I'm not too happy with the limits of `joint_3`, as in all other support pkgs we have specified the actual limits. I understand the linked kinematics problem, but I feel the joint limits in a xacro are not the place to work around that. For now we'll leave it like this though.

Testing was successful, so I'll wait on Travis to be happy and then I'll merge.

Thanks for the contribution :+1:

**Comment by gavanderhoorn** [comment]
We'll probably have to fix-up the collision meshes for `link_2` and possibly `link_4`, similar to what was reported for the basic 200iD (#34), but I'll do that later.

**Comment by gavanderhoorn** [comment]
Everything passed. Merging.

**Review by gavanderhoorn — APPROVED** [review]

**Pull request merged**
